annotation-server/src/common/cache-map.ts

```diff
@@ -0,0 +1,37 @@
+export abstract class CacheMap<Value, RetrieveFailedError> {
+    protected abstract retrieve(...keys: string[]): Promise<Value>;
+    protected abstract createError(...keys: string[]): RetrieveFailedError;
+    protected abstract validate(
+        valueOrError: Value | RetrieveFailedError,
+    ): Value;
+
+    cache: Map<string, Value | RetrieveFailedError>;
+
+    constructor() {
+        this.cache = new Map();
+    }
+
+    private mapKey(keys: string[]) {
```

Comment:
tldr; would complicate things

If we have the inheriting class implement this then it also has to implement de-joining. Let's say `GenePhenotypesByLookupkeyCache` uses keys that are a concatenation of symbol and lookupkey. Then its `retrieve` and `createError` methods would be called with the concatenation meaning it would have to de-concatenate to get the two strings it needs. This could be solved by concatenating with a separation character but I think it would be kind of ugly ;)